Updated Android Libs / Scripts for (arm-v8a (arm64) / x86_64 / x86 / arm-v7a) - Android NDK 23b Upgrade

danoli3 opened the pull request:
Android NDK 23.1 Requirement for Android Studio 4.2+ and new Build Systems

Updated most build scripts 

Android Features:
- C++17 support
- x86_64 Android Target
- NDK 23b - Allows for builds with Gradle Systems / Android.mk / CMake In Android Studio
- Update to Dependency Libraries
- Fixed a bunch of Android configure scripts for NDK 23 + and CMake option / make 

Apothecary Changes:
- Fixed some documentation 
- Added to some scripts SHA hash security to check if source zip has been modified / mostly build security for OpenSSL and Curl where these are highly important for clients auditing openFrameworks for security status 


- Updated **Assimp** from 4.0.1 to **5.1.6** (converted to CMake Config to Make)
- Updated **uriparser** from 0.8.5 to **0.9.6**
- Updated **pugixml** from 1.9 to **1.11.4**
- Updated **OpenSSL** from 1.1.0h to **1.1.1m** (+security patches)
- Updated **cURL** from 7.59.0 to **7.81.0** (fixed a bunch most platforms)
- Updated **OpenCV** from 4.0.1 to **4.5.5** - (fixed CMake build commands, emscripten, android, vs)
- Updated **tess2** - Android Build Script Fixed
- Updated **FreeImage** from 3.17.0 to **3.18.0** (iOS/tvOS/Android)
- Updated **libxml2** from 2.9.4 to **2.9.12** (Android CMake fixed )
- Updated **FreeType** from 2.7 to **2.11.1**
- Updated uri - (removed Boost from Android make, switched to CMake for Android)

- Removal: Android Boost Dependency (file system switched to C++17)
- Removal: Poco - Android Core (also fixed build script for Android if anyone wants to)

Other Notes:
- Fixed Android Install for NDK 23 and new download links 
- FreeImage source code Updated to 3.18.0 for Android and patched for C++17
- Removed FreeImage SubComponents (OpenEXR / LibRAWlite) removed from FreeImage util they update to support new compile targets. Updated github.com/danoli3/FreeImage to support C++17 and had to remove these broken libraries for now.
- Fixed iOS Configure targets for tvOS / bitcode 
- Updated Docker to use CMake  3.22.1
- Fixed **linuxarmv7l/install.sh** for new URL

Fixed other scripts in the process:
- Updated **pixman** from 0.34.0 to **0.40.0**
- Updated **pkg-config.sh** from 0.28 to **0.29.2**
- Updated **Cairo** from 1.14.12 to **1.17.4**


Current Issues:

ofTheo commented:
@danoli3 can you use your repo to do the PRs instead of a branch on the OF repo? 
Also if you could delete the `build-android-arm64` branch from this repo ( or I can do it if you like? ) that would be awesome.

danoli3 commented:
@ofTheo always was

> @danoli3 can you use your repo to do the PRs instead of a branch on the OF repo?
> Also if you could delete the `build-android-arm64` branch from this repo ( or I can do it if you like? ) that would be awesome.

Always was on my fork, not sure how it got pushed to upstream -- not on there just checked or in my local system to have that as a remote at all so not sure how 
I think I saw this happen on main openframeworks branch as well once but 
<img width="561" alt="Screen Shot 2022-01-09 at 3 00 30 pm" src="https://user-images.githubusercontent.com/830748/148668872-55f9089e-f1a8-4a7a-b737-3e2fb88ab1fc.png">
also I do not have upstream at all set as a remote, only my danoli3/

danoli3 commented:
Remaining Issue with linking OpenSSL to ... missing something in LDFlags it would seem
```
ld: error: cannot open crtbegin_so.o: No such file or directory
ld: error: cannot open crtend_so.o: No such file or directory
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [libcrypto.so] Error 1
```

danoli3 commented:
OpenSSL 1.1.1m Fixed!

danoli3 commented:
Almost there now, just keep catching issues with dependancies and pushing everything to bleeding edge

danoli3 commented:
Okay for some reason there is no way Cairo ever worked, FreeType binding was not working prior based on logs on past automation : Fixed and updated for macOS 

I think that should fix remaining automation issues

danoli3 commented:
So close!!! Only windows vs left to have a look what I broke by updating in automation other than that! Super close here

ofTheo commented:
👏 Wow - awesome! 
I can take a look at the VS stuff later today if its helpful. 

Hold off on merging too as I just want to go through all the changes first. 
Looks great though!!

danzeeeman commented:
RE Windows we gotta update to VS2022 for the docs and show people how to
download the VS217 cpp libs or just update everything to VS2022

On Tue, Jan 25, 2022 at 11:45 AM Theodore Watson ***@***.***>
wrote:

> 👏 Wow - awesome!
> I can take a look at the VS stuff later today if its helpful.
>
> Hold off on merging too as I just want to go through all the changes first.
> Looks great though!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/apothecary/pull/201#issuecomment-1021389914>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFPVTBCXFUDPSFORQ32TB3UX3HQXANCNFSM42SB6J3Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
***@***.***

ofTheo commented:
@danzeeeman did this yesterday! 

Download now mentions 2017-2022 and setup instructions now link to 2019. Need to update the installer image and then it should be good.  
https://openframeworks.cc/download/ 
https://openframeworks.cc/setup/vs/

Thinking also it might be time to drop Windows 32bit. 
Would slim things down a lot and pretty much all other platforms are 64bit only.

danoli3 commented:
Yeah I don't really think win32 windows used anymore, more ARM64 windows

danzeeeman commented:
Dump it!

On Tue, Jan 25, 2022 at 7:48 PM Dan Rosser ***@***.***> wrote:

> Yeah I don't really think win32 windows used anymore, more ARM64 windows
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/apothecary/pull/201#issuecomment-1021747796>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFPVTEMXDOC5Y5YUILHFVLUX5AHTANCNFSM42SB6J3Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Sent from Gmail Mobile #autocorrect

danoli3 commented:
iOS/tvOS Make sure to Add Security.framework when testing new libs if that's not there already

ofTheo commented:
Ah, so close! 

Looking at the last error I think you could set these vars:

`PYTHON_VERSION_STRING=3.9.10`
`PYTHON_EXECUTABLE=C:/hostedtoolcache/windows/Python/3.9.10/x64/python.exe`

Then the exe would match the libs which it finds.

danoli3 commented:
Strange I disabled python bindings and still an error need to look deeper and get that log file. Really need to get this building on a Microsoft machine locally

danoli3 commented:
Actually be great to get CUDA later for Windows OpenCV! https://thinkinfi.com/install-opencv-gpu-with-cuda-for-windows-10/

danzeeeman commented:
+1 on cuda openCV

On Mon Mar 14, 2022, 12:18 AM GMT, Dan Rosser ***@***.***>
wrote:

Actually be great to get CUDA later for Windows OpenCV!
https://thinkinfi.com/install-opencv-gpu-with-cuda-for-windows-10/

—
Reply to this email directly, view it on GitHub
<https://github.com/openframeworks/apothecary/pull/201#issuecomment-1066216856>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AAFPVTDXGD3SMAROMKKL5GLU72AVVANCNFSM42SB6J3Q>
.
Triage notifications on the go with GitHub Mobile for iOS
<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
or Android
<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

You are receiving this because you were mentioned.Message ID:
***@***.***>

ofTheo commented:
I've got a Windows PC here and will try with your branch. 
I think this could be the answer ( assuming Python is the same one Cmake is detecting ).

`python -mpip install numpy`
https://github.com/numpy/numpy/issues/15424

ofTheo commented:
hmm actually ignore the above. 
the issue I think is not numpy or python, but cmake not finding the compiler. 

```
CMake Error at cmake/OpenCVCompilerOptimizations.cmake:593 (message):
  Compiler doesn't support baseline optimization flags:
```

ofTheo commented:
fyi - tried a bunch of stuff here and nothing seems closer:
https://github.com/ofTheo/apothecary/commits/vs-fix-libs 

still not sure if the numpy error is related - but trying to get it to go away....

danoli3 commented:
Looks like we are not alone in this issue https://github.com/astropy/pytest-doctestplus/issues/177

Lets move to windows-2019 for environment looks like it has the missing compiler for numcpy. 

Furthermore since March 2022 GitHub Actions upgraded to MS environment significantly in windows-latest, should look at that soon

danoli3 commented:
Ah yes! so windows-2016 is gone! no wonder ! Removed March 15

danoli3 commented:
Okay sweet, added a lot of the vs2019 matrix upgrades for CMAKE build types / VS 16 2019 and VS 2022 formatting 

Pretty close I think

danoli3 commented:
arm error: The unauthenticated git protocol on port 9418 is no longer supported (must change git:// to https:// for github repos

ofTheo commented:
@danoli3 what's left on this to get merged? 
I should have time next week to jump into apothecary and CI stuff so that we can try hosting the compiled libs on a nightly release tag.

danoli3 commented:
Current Issue: 
- [ ] linuxarm6 compiling libxml2 / re-configuration of linux arm system updated 

```
Building "libxml2"
[248](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:248)

[249](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:249)
-- The C compiler identification is unknown
[250](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:250)
CMake Error at CMakeLists.txt:13 (project):
[251](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:251)
  The CMAKE_C_COMPILER:
[252](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:252)

[253](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:253)
    /home/runner/work/apothecary/apothecary/scripts/linuxarmv6l/rpi_toolchain/bin/arm-linux-gnueabihf-gcc
[254](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:254)

[255](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:255)
  is not a full path to an existing compiler tool.
[256](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:256)

[257](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:257)
  Tell CMake where to find the compiler by setting either the environment
[258](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:258)
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
[259](https://github.com/openframeworks/apothecary/runs/6318082837?check_suite_focus=true#step:5:259)
  the compiler, or to the compiler name if it is in the PATH.
  ```
  
  
  Windows ARM64 :

danoli3 commented:
SVGTiny having issues with libxml2 - using system libs.


```
In file included from libdom/bindings/xml/libxml_xmlparser.c:12:
[536](https://github.com/openframeworks/apothecary/runs/6398271015?check_suite_focus=true#step:6:536)
In file included from /Users/runner/work/apothecary/apothecary/scripts/../out/libxml2/include/libxml/parser.h:812:
[537](https://github.com/openframeworks/apothecary/runs/6398271015?check_suite_focus=true#step:6:537)
/Users/runner/work/apothecary/apothecary/scripts/../out/libxml2/include/libxml/encoding.h:31:10: fatal error: 'unicode/ucnv.h' file not found
[538](https://github.com/openframeworks/apothecary/runs/6398271015?check_suite_focus=true#step:6:538)
#include <unicode/ucnv.h>
[539](https://github.com/openframeworks/apothecary/runs/6398271015?check_suite_focus=true#step:6:539)
         ^~~~~~~~~~~~~~~~
[540](https://github.com/openframeworks/apothecary/runs/6398271015?check_suite_focus=true#step:6:540)
/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -c libparserutils/src/charset/aliases.c -o libparserutils/src/charset/aliases.o -O2 -arch x86_64  -isysroot /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator15.2.sdk -pipe -Oz -gdwarf-2 -fembed-bitcode -fPIC -mios-simulator-version-min=10.0 -I/Users/runner/work/apothecary/apothecary/scripts/../out/libxml2/include  -D_BSD_SOURCE -D_DEFAULT_SOURCE -DWITHOUT_ICONV_FILTER -fPIC -I/Users/runner/work/apothecary/apothecary/apothecary/build/svgtiny/include/ -I/Users/runner/work/apothecary/apothecary/apothecary/build/svgtiny/src -Wall -W -Wundef -Wpointer-arith -Wwrite-strings -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -std=c99 -Ilibdom/include -Ilibdom/src -Ilibdom/bindings -Ilibwapcaplet/include -Ilibparserutils/include -Ilibparserutils/src
```

On my local it is compiling: not sure why the ./out/ ?
```
-I/Users/runner/work/apothecary/apothecary/scripts/../out/libxml2/include
```
